README.rst
drewtilities
=============

| |BSD3 License|

| |Build Status|

| |Coverage Status|

| |Issue Count|

.. image:: https://badge.fury.io/py/drewtilities.svg
    :target: https://badge.fury.io/py/drewtilities

.. image:: https://badge.waffle.io/alixnovosi/drewtilities.png?label=ready&title=Ready
    :target: https://waffle.io/alixnovosi/drewtilities
    :alt: 'Stories in Ready'

.. |BSD3 License| image:: http://img.shields.io/badge/license-BSD3-brightgreen.svg
   :target: https://tldrlegal.com/license/bsd-3-clause-license-%28revised%29
.. |Build Status| image:: https://travis-ci.org/alixnovosi/drewtilities.svg?branch=master
   :target: https://travis-ci.org/alixnovosi/drewtilities
.. |Coverage Status| image:: https://coveralls.io/repos/alixnovosi/drewtilities/badge.svg?branch=master&service=github
   :target: https://coveralls.io/github/alixnovosi/drewtilities?branch=master
.. |Issue Count| image:: https://codeclimate.com/github/alixnovosi/drewtilities/badges/issue_count.svg
   :target: https://codeclimate.com/github/alixnovosi/drewtilities

=====
short
=====
Utility functions, created by Drew, for Drew.
Used in `botskeleton`_ (and all the bots that use it) and `puckfetcher`_.

.. _botskeleton: https://github.com/alixnovosi/botskeleton
.. _puckfetcher: https://github.com/alixnovosi/puckfetcher

====
long
====
These are in the order defined in :code:`drewtilities.py`,
which in turn is basically defined by order of creation.
Not really organized nicely,
sorry.

=======
Methods
=======

-----------------------------
:code:`ensure_dir(directory)`
-----------------------------
Create directory if it doesn't exist.

-------------------------
:code:`expand(directory)`
-------------------------
Expand :code:`~` to :code:`$HOME`,
and expand other environment variables to their values,
in the provided path.
Designed for UNIX-based platforms:
unclear what will happen on others.

-----------------------------------------------------------
:code:`generate_downloader(headers, args, max_per_hour=30)`
-----------------------------------------------------------
Produce a callable downloader.
Provide headers for URL call,
and provide some args
(for the rate limiting
(see that later in the document)),
which will be used to identify this function in rate-limiting.
:code:`max_per_hour` is optional:
use it to decide how many times per hour downloads can happen.
The default is 30 times per hour.
The returned downloader is called as follows:

.. code-block:: python

    downloader = generate_downloader({"foo": "bar"}, "an identifier string")
    downloader(url=url, dest=dest)

where :code:`url` is where to download FROM and :code:`dest` is where to SAVE the file.
It will make the directory for the file if it does not exist.
A text progress bar is shown during download,
and while rate limit is blocking.

-----------------------------
:code:`max_clamp(val, limit)`
-----------------------------
Return :code:`val` if it is less than :code:`limit`,
otherwise return :code:`limit`.

------------------------------------
:code:`parse_int_string(int_string)`
------------------------------------
Given a string like "1 23 4-8 32 1",
return a unique list of those integers in the string,
and the integers in the ranges in the string.
For this example,
the output would be [1, 4, 5, 6, 7, 8, 23, 32].
Non-numbers ignored.
Not necessarily sorted

-----------------------------------------
:code:`rate_limited(max_per_hour, *args)`
-----------------------------------------
A decorator to rate-limit a function
(ensures that it runs no more than :code:`max_per_hour` times per hour by sleeping sometimes).
Called like this:

.. code-block:: python

    @util.rate_limited(120, name)
    def a_function(var1, var2="foo"):

-----------------------------------------
:code:`sanitize(filename, platform=None)`
-----------------------------------------
Remove disallowed characters from potential filename.
Provide platform to sanitize for that platform,
otherwise current platform is found and used.

-------------------------------------------------------
:code:`set_up_logging(log_filename="log", verbosity=0)`
-------------------------------------------------------
Set up a logger with reasonable settings.
Return it for log calls.

--------------------------------------------------------
:code:`random_line(file_path, encoding=FORCED_ENCODING)`
--------------------------------------------------------
At time of writing,
:code:`FORCED_ENCODING` is "UTF-8".
Refer to code for latest.
Return a line from the provided file at random.


CHANGELOG
1.3.2 (2019-05-30)
* BUGFIX
    * forgot to change S to D after testing for log rotation.

1.3.1 (2019-05-30)
* BUGFIX
    * minor type fixes.

1.3.0 (2019-05-30)
* FEATURE
    * optional date-based rotation instead of file size rotation.

* BUGFIX
    * use UTC timestamps and clean up timestamp format to show that
    * not considered a breaking change because timestamps were ambiguous before.

1.2.2 (2019-03-05)
1.2.1
* BUGFIX
    * Fix format string where we printed remaining instead of the remaining amount of time.

* DEV
    * Two updates to force a dependency update

1.2.0 (2018-10-15)
* FEATURE
    * sanitize supports sanitizing for windows filenames.
      disallowed characters are mapped to allowed characters.
      see drewtilities/drewtilities.py for rules.
      all other platforms only sanitize slashes to hyphens

1.1.2 (2018-09-24)
Another forced update to get README fix through.

1.1.1 (2018-09-24)
Forced update to get README fix through.

1.1.0 (2018-09-24)
change download rate-limit, and some other cleanups
* FEATURE
    * generate_downloader will now let you provide a max_per_hour for its internal rate-limit
      default is 30
    * provide a nice README that describes the various functions provided here
* DEV
    * setup.py cleanups and new version pins
    * make CHUNK_SIZE a variable

1.0.5 (2017-12-05)
Test fix.
* DEV
    * fix typo'd name in test

1.0.4 (2017-11-26)
various fixes
* DEV
    * fix travis.yml to fix install
    * VERSION needed in package_data
    * package needs to be drewtilities, not drewtils, to be on PyPI

1.0.3 (2017-11-26)
no change, forced update for test purposes

1.0.2 (2017-11-25)
remove VERSION from deploy
* DEV
    * We don't need to put VERSION in package_data

1.0.1 (2017-11-25)
initial migration
* FEATURES
    * initial migration from puckfetcher codebase
* DEV
    * add basic testing and CI setup


drewtils/test/test_drewtils.py
"""Tests for the util module."""
import timeit

import requests

import drewtils as util

TEST_URL = "https://raw.githubusercontent.com/andrewmichaud/drewtils/master/LICENSE"


def test_parse_int_string() -> None:
    """Test int string parser."""
    doc_example = "1 23 4-8 32 1"
    res = util.parse_int_string(doc_example)
    assert sorted(res) == [1, 4, 5, 6, 7, 8, 23, 32]

    invalid_nums = "1 fish a 4 1 23-- 5 5 5 5 5 9"
    res = util.parse_int_string(invalid_nums)
    assert sorted(res) == [1, 4, 5, 9]

    ranges = "1-5, 7-10, 9-12"
    res = util.parse_int_string(ranges)
    assert sorted(res) == [1, 2, 3, 4, 5, 7, 8, 9, 10, 11, 12]


def test_rate_limiting() -> None:
    """Test rate limiting (with a short limit, to not make tests drag on)."""

    @util.rate_limited(3600, "test")
    def _test_helper() -> None:
        requests.get(TEST_URL)

    _test_helper()

    res = timeit.timeit(_test_helper, number=4)
    assert abs(res - 4) < 1.0


drewtilities/test/test_drewtilities.py
"""Tests for the util module."""
import timeit

import requests

import drewtilities as util

TEST_URL = "https://raw.githubusercontent.com/andrewmichaud/drewtilities/master/LICENSE"


def test_parse_int_string() -> None:
    """Test int string parser."""
    doc_example = "1 23 4-8 32 1"
    res = util.parse_int_string(doc_example)
    assert sorted(res) == [1, 4, 5, 6, 7, 8, 23, 32]

    invalid_nums = "1 fish a 4 1 23-- 5 5 5 5 5 9"
    res = util.parse_int_string(invalid_nums)
    assert sorted(res) == [1, 4, 5, 9]

    ranges = "1-5, 7-10, 9-12"
    res = util.parse_int_string(ranges)
    assert sorted(res) == [1, 2, 3, 4, 5, 7, 8, 9, 10, 11, 12]


def test_rate_limiting() -> None:
    """Test rate limiting (with a short limit, to not make tests drag on)."""

    @util.rate_limited(3600, "test")
    def _test_helper() -> None:
        requests.get(TEST_URL)

    _test_helper()

    res = timeit.timeit(_test_helper, number=4)
    assert abs(res - 4) < 1.0

def test_sanitize() -> None:
    """Test sanitize on multiple platforms."""

    test_filename = "asdfadfasdf"
    assert test_filename == util.sanitize(test_filename)

    # test posix replace
    test_filename = "yoyoyo////"
    expected = "yoyoyo----"
    assert util.sanitize(test_filename, "linux") == expected

    # test windows replace
    test_filename = "a\\a:a*a?a\"a<a>a|a"
    expected = "aÿa÷a¤a¿a¨a«a»a¦a"
    assert util.sanitize(test_filename, "win32") == expected

    # and test windows replace only happens when demanded
    assert util.sanitize(test_filename, "asdf") == test_filename


drewtilities/drewtilities.py
"""Utility methods for Python packages (written with puckfetcher and botskeleton in mind)."""
import logging
from logging.handlers import RotatingFileHandler, TimedRotatingFileHandler
import math
import os
import random
import sys
import time
from typing import Any, Callable, Dict, List, Set

import requests

import clint.textui as tui

HERE = os.path.abspath(os.path.dirname(__file__))
LOG = logging.getLogger("root")

CHUNK_SIZE=1024
FORCED_ENCODING = "UTF-8"
LAST_CALLED: Dict[str, float] = {}

def ensure_dir(directory: str) -> None:
    """Create a directory if it doesn't exist."""
    if not os.path.isdir(directory):
        LOG.debug(f"Directory {directory} does not exist, creating it.")
        os.makedirs(directory)

def expand(directory: str) -> str:
    """Apply expanduser and expandvars to directory to expand '~' and env vars."""
    temp1 = os.path.expanduser(directory)
    return os.path.expandvars(temp1)

def generate_downloader(headers: Dict[str, str], args: Any, max_per_hour: int=30
                        ) -> Callable[..., None]:
    """Create function to download with rate limiting and text progress."""

    def _downloader(url: str, dest: str) -> None:

        @rate_limited(max_per_hour, args)
        def _rate_limited_download() -> None:

            # Create parent directory of file, and its parents, if they don't exist.
            parent = os.path.dirname(dest)
            if not os.path.exists(parent):
                os.makedirs(parent)

            response = requests.get(url, headers=headers, stream=True)
            LOG.info(f"Downloading from '{url}'.")
            LOG.info(f"Trying to save to '{dest}'.")

            length = response.headers.get("content-length")
            if length is None:
                total_length = 0
            else:
                total_length = int(length)

            expected_size = (total_length / CHUNK_SIZE) + 1
            chunks = response.iter_content(chunk_size=CHUNK_SIZE)

            open(dest, "a", encoding=FORCED_ENCODING).close()
            # per http://stackoverflow.com/a/20943461
            with open(dest, "wb") as stream:
                for chunk in tui.progress.bar(chunks, expected_size=expected_size):
                    if not chunk:
                        return
                    stream.write(chunk)
                    stream.flush()

        _rate_limited_download()

    return _downloader

def max_clamp(val: int, limit: int) -> int:
    """Clamp int to limit."""
    return min(val, limit)

def parse_int_string(int_string: str) -> List[int]:
    """
    Given a string like "1 23 4-8 32 1", return a unique list of those integers in the string and
    the integers in the ranges in the string.
    Non-numbers ignored. Not necessarily sorted
    """
    cleaned = " ".join(int_string.strip().split())
    cleaned = cleaned.replace(" - ", "-")
    cleaned = cleaned.replace(",", " ")

    tokens = cleaned.split(" ")
    indices: Set[int] = set()
    for token in tokens:
        if "-" in token:
            endpoints = token.split("-")
            if len(endpoints) != 2:
                LOG.info(f"Dropping '{token}' as invalid - weird range.")
                continue

            start = int(endpoints[0])
            end = int(endpoints[1]) + 1

            indices = indices.union(indices, set(range(start, end)))

        else:
            try:
                indices.add(int(token))
            except ValueError:
                LOG.info(f"Dropping '{token}' as invalid - not an int.")

    return list(indices)

# Modified from https://stackoverflow.com/a/667706
def rate_limited(max_per_hour: int, *args: Any) -> Callable[..., Any]:
    """Decorator to limit function to N calls/hour."""
    min_interval = 3600.0 / float(max_per_hour)

    def _decorate(func: Callable[..., Any]) -> Callable[..., Any]:
        things = [func.__name__]
        things.extend(args)
        key = "".join(things)
        LOG.debug(f"Rate limiter called for '{key}'.")
        if key not in LAST_CALLED:
            LOG.debug(f"Initializing entry for '{key}'.")
            LAST_CALLED[key] = 0.0

        def _rate_limited_function(*args: Any, **kargs: Any) -> Any:
            last_called = LAST_CALLED[key]
            now = time.time()
            elapsed = now - last_called
            remaining = min_interval - elapsed
            LOG.debug(f"Rate limiter last called for '{key}' at {last_called}.")
            LOG.debug(f"Remaining cooldown time for '{key}' is {remaining}.")

            if remaining > 0 and last_called > 0.0:
                LOG.info(f"Self-enforced rate limit hit, sleeping {remaining} seconds.")
                for i in tui.progress.bar(range(math.ceil(remaining))):
                    time.sleep(1)

            LAST_CALLED[key] = time.time()
            ret = func(*args, **kargs)
            LOG.debug(f"Updating rate limiter last called for '{key}' to {now}.")
            return ret

        return _rate_limited_function
    return _decorate

def sanitize(filename: str, platform: str=None) -> str:
    """
    Remove disallowed characters from potential filename. Currently only guaranteed on Linux and
    OS X.
    """
    win_map = {
        # taken from samba Catia module.
        # https://www.samba.org/samba/docs/current/man-html/vfs_catia.8.html
        "\"": "¨",
        "*": "¤",
        "/": "ÿ",
        ":": "÷",
        "<": "«",
        ">": "»",
        "?": "¿",
        "\\": "ÿ",
        "|": "¦",
    }

    posix_map = {
        "/": "-",
    }

    if platform is None:
        platform = sys.platform

    if platform.startswith("win32"):
        replace_map = win_map
    else:
        replace_map = posix_map

    for key, entry in replace_map.items():
        filename = filename.replace(key, entry)

    return filename

def set_up_logging(
        *,
        log_filename: str = "log",
        verbosity: int = 0,
        use_date_logging: bool = False,
) ->logging.Logger:
    """Set up proper logging."""

    # log everything verbosely
    LOG.setLevel(logging.DEBUG)

    logging.Formatter.converter = time.gmtime

    handler: Any
    if use_date_logging:
        handler = TimedRotatingFileHandler(
            filename=log_filename,
            when="S",
            utc=True,
        )

    else:
        handler = RotatingFileHandler(
            filename=log_filename,
            maxBytes=1024000000,
            backupCount=10,
        )

    formatter = logging.Formatter(
        fmt="%(asctime)s.%(msecs)03dZ - %(levelname)s - %(module)s - %(message)s",
        datefmt="%Y-%m-%dT%H:%M:%S",
    )

    handler.setFormatter(formatter)
    handler.setLevel(logging.DEBUG)
    LOG.addHandler(handler)

    # Provide a stdout handler logging at INFO.
    stream_handler = logging.StreamHandler(sys.stdout)
    simple_form = logging.Formatter(fmt="%(message)s")
    stream_handler.setFormatter(simple_form)

    if verbosity > 0:
        stream_handler.setLevel(logging.DEBUG)
    else:
        stream_handler.setLevel(logging.INFO)

    LOG.addHandler(stream_handler)

    return LOG

def random_line(file_path: str, encoding: str = FORCED_ENCODING) -> str:
    """Get random line from a file."""
    # Fancy alg from http://stackoverflow.com/a/35579149 to avoid loading full file.
    line_num = 0
    selected_line = ""
    with open(file_path, encoding=encoding) as stream:
        while True:
            line = stream.readline()
            if not line:
                break
            line_num += 1
            if random.uniform(0, line_num) < 1:
                selected_line = line

    return selected_line.strip()
